prow/cmd/artifact-uploader/main.go

```diff
@@ -0,0 +1,166 @@
+		return errors.New("number of workers cannot be zero")
+	}
+
+	if o.ProwJobNamespace == "" {
+		return errors.New("namespace containing ProwJobs not configured")
+	}
+
+	return o.Options.Validate()
+}
+
+const (
+	// JSONConfigEnvVar is the environment variable that
+	// utilities expect to find a full JSON configuration
+	// in when run.
+	JSONConfigEnvVar = "ARTIFACTUPLOAD_OPTIONS"
+)
+
+// ConfigVar exposes the environment variable used
+// to store serialized configuration
+func (o *Options) ConfigVar() string {
+	return JSONConfigEnvVar
+}
+
+// LoadConfig loads options from serialized config
+func (o *Options) LoadConfig(config string) error {
+	return json.Unmarshal([]byte(config), o)
+}
+
+// BindOptions binds flags to options
+func (o *Options) BindOptions(flags *flag.FlagSet) {
+	flags.IntVar(&o.NumWorkers, "num-workers", 25, "Number of threads to use for processing updates.")
+	flags.StringVar(&o.ProwJobNamespace, "prow-job-ns", "", "Namespace containing ProwJobs.")
+	gcsupload.BindOptions(o.Options, flags)
+}
+
+// Complete internalizes command line arguments
+func (o *Options) Complete(args []string) {
+	o.Options.Complete(args)
+}
+
+// loadClusterConfig loads connection configuration
+// for the cluster we're deploying to. We prefer to
+// use in-cluster configuration if possible, but will
+// fall back to using default rules otherwise.
+func loadClusterConfig() (*rest.Config, error) {
+	clusterConfig, err := rest.InClusterConfig()
+	if err == nil {
+		return clusterConfig, nil
+	}
+
+	credentials, err := clientcmd.NewDefaultClientConfigLoadingRules().Load()
+	if err != nil {
+		return nil, fmt.Errorf("could not load credentials from config: %v", err)
+	}
+
+	clusterConfig, err = clientcmd.NewDefaultClientConfig(*credentials, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not load client configuration: %v", err)
+	}
+	return clusterConfig, nil
+}
+
+func (o *Options) Run() error {
+	clusterConfig, err := loadClusterConfig()
+	if err != nil {
+		return fmt.Errorf("failed to load cluster config: %v", err)
+	}
+
+	client, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		return err
+	}
+
+	prowJobClient, err := kube.NewClientInCluster(o.ProwJobNamespace)
+	if err != nil {
+		return err
+	}
+
+	controller := artifact_uploader.NewController(client.CoreV1(), prowJobClient, o.Options)
+
+	stop := make(chan struct{})
+	defer close(stop)
+	go controller.Run(o.NumWorkers, stop)
+
+	// Wait forever
+	select {}
+}
+
+func main() {
+	o := NewOptions()
+	if err := options.Load(o); err != nil {
+		logrus.Fatalf("Could not resolve options: %v", err)
+	}
+
+	if err := o.Validate(); err != nil {
+		logrus.Fatalf("Invalid options: %v", err)
+	}
+
+	logrus.SetFormatter(
+		logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "artifact-uploader"}),
+	)
+
+	if err := o.Run(); err != nil {
+		logrus.WithError(err).Fatal("Failed to push artifacts to GCS")
```

Comment:
This error message needs to be something different. Feels more like a "Failed to run the artifacts uploader"